proxy/grpcproxy/cluster.go

```diff
@@ -15,20 +15,122 @@
 package grpcproxy
 
 import (
+	"fmt"
+	"sync"
+
 	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/naming"
 	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
 
 	"golang.org/x/net/context"
+	"golang.org/x/time/rate"
+	"google.golang.org/grpc"
+	gnaming "google.golang.org/grpc/naming"
 )
 
 type clusterProxy struct {
 	client *clientv3.Client
+
+	// self is true to return its own client listen URLs
+	// for member list API
+	self  bool
+	laddr string
+
+	wa   gnaming.Watcher
+	umu  sync.RWMutex
+	umap map[string]struct{}
 }
 
-func NewClusterProxy(c *clientv3.Client) pb.ClusterServer {
-	return &clusterProxy{
+// NewClusterProxy takes optional prefix to fetch grpc-proxy member endpoints.
+// 'self' is true then member list API returns the client listen addresses.
+// The returned channel is closed when there is grpc-proxy endpoint registered
+// and the client's context is canceled so the 'register' loop returns.
+func NewClusterProxy(c *clientv3.Client, self bool, laddr string, prefix string, advaddr string, ttl int) (pb.ClusterServer, <-chan struct{}, error) {
```

Comment:
> This isn't a good interface; the return results are too dependent on the arguments. Just have donec closed when all goroutines associated with the cluster proxy are shutdown (like the other proxy services). Right now it will need special code for donec based on whether a prefix was given or not.

Now it returns `donec` as other proxy servers, but the arguments got too big... Any suggestion? Thanks.